### DISCUSSION TITLE
SSR Question: How to pass data (e.g. breadcrumbs) from pages to components (e.g. footer) defined in layout?

### DISCUSSION THREAD

**Discussion by AlexBa** [nuxt/nuxt#19424] [Questions] [answered]
Hello all,
I've been dealing with a problem I just have for hours and have searched half the internet :)

I would like to render a blog article on a page and put the breadcrumbs dynamically into the footer. For the client side it's very easy with `useState` . But the server side part is not really working. The initial HTML just don't render the breadcrumbs.

By the way: I have the same problem with header language switch, which is based on the dynamic translated slug of the blog post.

Here is my live example: https://www.shop-studio.io/blog/shopware-vs-woocommerce-what-is-the-better-solution

Here is my code. First some helper functions: 
```
import { Ref } from '@vue/reactivity'

interface Breadcrumb {
  name: string | null
  position: number
  route: object | null
}

export const getBreadcrumbs = (): Ref<Breadcrumb[]> => {
  return useState<Breadcrumb[]>('breadcrumbs', () => [])
}

export const useBreadcrumbs = (breadcrumbs: Breadcrumb[]): void => {
  const state = getBreadcrumbs()

  state.value = breadcrumbs
}
```

My code from pages/blog/[slug.vue]:
```
<script setup lang="ts">
...

const blogPost = await getSingle('blog-posts', slug, 'blog-slug')

if (blogPost) {
 useBreadcrumbs([
    {
      name: 'Blog',
      position: 1,
      route: { name: 'blog' },
    },
    {
      name: blogPost.metadata.title,
      position: 2,
      route: null,
    },
  ])
}

...
</script>
```

My code from components/footer.vue:
```
<template>
  <footer>
     <template v-if="breadcrumbItems.length > 0">
            <div class="row">
              <div class="col-12">
                <div class="footer__top">
                  <breadcrumbs :items="breadcrumbItems" class="mb-4" />
                </div>
              </div>
            </div>
            <div class="row">
              <div class="col-12">
                <hr class="footer_separator mb-5" />
              </div>
            </div>
        </template>
    </footer>
</template>

<script setup lang="ts">
...

const breadcrumbItems = getBreadcrumbs()
// breadcrumbItems.value is empty ("[]"). But only on SSR.

...
</script>
```

Additionally, I tried out definePageMeta as well, but it's not working with async data (build macro).

My question is: Am I'm doing something wrong? What is the solution to this problem? I tested out Nuxt 3.2.3 and 3.0.0.

Thank you in advance!

**Comment by AlexBa** [comment]
By the way: This question is very similar: https://github.com/nuxt/nuxt/discussions/18451

**Comment by TheAlexLichter** [comment]
TL;DR - you cannot.

Long answer (adapted from https://github.com/nuxt/nuxt/issues/15474#issuecomment-1397380967):

You cannot alter components "higher up" like a layout or `app.vue` when rendering a page component as they are already rendered during SSR (top-down). Same applies when rendering components "further down the tree" which change values that should be reflected "further up" (e.g. via Pinia or Composables).

**Reply by AlexBa** [reply]
Thank you for your answer! :)

**Reply by akarelas** [reply]
It would be nice if it were possible though, since maou-shonen's solution would make the layout's setup function be executed on every route change

**Comment by maou-shonen** [comment, marked as answer]
Here is a solution. Move <NuxtLayout> from app.vue to pages.
The benefit of this approach is that you can easily pass props to the layout without having to create a lot of state.
```diff
// app.vue
<template>
--- <NuxtLayout>
      <NuxtPage />
--- </NuxtLayout>
</template>
```
```diff
// pages/index.vue
<template>
+++  <NuxtLayout>
    <div>
      <content>
    </div>
+++  </NuxtLayout>
</template>
```

A more radical solution <- I adopted this solution:
Move all layouts to components.
The downside is that you can't use the layout with definePageMeta and setLayout (use `component is`),
but incorrect prop passing can be checked by linting tools.
```bash
mv layouts/default.vue components/layout/Base.vue
```
```diff
// app.vue
<template>
--- <NuxtLayout>
      <NuxtPage />
--- </NuxtLayout>
</template>
```
```diff
// pages/index.vue
<template>
+++  <LayoutBase>
    <div>
      <content>
    </div>
+++  </LayoutBase>
</template>
```

**Reply by AlexBa** [reply]
Thank you for your solution. I had something similar in mind, but somehow I thought there might exist a better solution.

**Reply by Goryyn** [reply]
Can I see your code?

**Reply by akarelas** [reply]
This has the problem that the setup function of the layout will be executed on every route change.

It would be nice, therefore, if there was a mechanism for a page to pass data to the layout.
